### PR TITLE
Update E2E Tutorial w/ vLLM and HF Hub

### PR DESCRIPTION
**Pull request by joecummings**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog
What are the changes made in this PR?
*

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [ ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

File: docs/source/tutorials/e2e_flow.rst [hunk 1]
```diff
@@ -25,222 +25,239 @@ from the community seemlessly with torchtune.
...
+    Optimizer and loss are initialized.
+    Loss is initialized.
+    Dataset and Sampler are initialized.
+    Learning rate scheduler is initialized.
+    Profiling disabled.
+    Profiler config after instantiation: {'enabled': False}
+    1|3|Loss: 1.943998098373413:   0%|                    | 3/1617 [00:21<3:04:47,  6.87s/it]
+
+Congrats on training your model! Let's take a look at the artifacts produced by torchtune. A simple way of doing this is by running `tree -a path/to/outputdir`, which should show something like the tree below.
+There are 4 types of folders:
```
Comment (felipemello1) [inline docs/source/tutorials/e2e_flow.rst:130]: 3 folders

File: docs/source/tutorials/e2e_flow.rst [hunk 2]
```diff
@@ -25,222 +25,239 @@ from the community seemlessly with torchtune.
...
+
+1) **recipe_state**: Holds recipe_state.pt with the information necessary to restart the last intermediate epoch. For more information, please check our deep-dive :ref:`Checkpointing in torchtune <understand_checkpointer>`.;
+2) **logs**: Defined in your config in metric_logger;
+3) **epoch_{}**: Contains your new trained model weights plus all original files of the model, except the checkpoints, making it easy for you to choose an specific epoch to run inference on or push to a model hub;
```
Comment (felipemello1) [inline docs/source/tutorials/e2e_flow.rst:134]: i made some changes to it in checkpointer, based on evans comments. Probably work copying/pasting it here.

File: docs/source/tutorials/e2e_flow.rst [hunk 3]
```diff
@@ -25,222 +25,239 @@ from the community seemlessly with torchtune.
...
+    │   ├── original_repo_id.json
+    │   ├── README.md
+    │   ├── special_tokens_map.json
+    │   ├── tokenizer_config.json
+    │   ├── tokenizer.json
+    │   └── USE_POLICY.md
+    ├── logs
+    │   └── log_1734652101.txt
+    └── recipe_state
+        └── recipe_state.pt
```
Comment (felipemello1) [inline docs/source/tutorials/e2e_flow.rst:163]: if we trained for only 1 epoch, there shouldnt be a recipe state. You should probably keep epoch_0 and epoch_1

File: docs/source/tutorials/e2e_flow.rst [hunk 4]
```diff
@@ -25,222 +25,239 @@ from the community seemlessly with torchtune.
...
+    batch_size: 8
+    enable_kv_cache: True
 
-Now, let's run the recipe.
-
-.. code-block:: bash
-
-    tune run eleuther_eval --config ./custom_eval_config.yaml
+    # Quantization specific args
+    quantizer: null
```
Comment (felipemello1) [inline docs/source/tutorials/e2e_flow.rst:245]: evan suggested to remove all of that

File: docs/source/tutorials/e2e_flow.rst [hunk 5]
```diff
@@ -25,222 +25,239 @@ from the community seemlessly with torchtune.
...
-    [eleuther_eval.py:195] Eval completed in 121.27 seconds.
-    [eleuther_eval.py:197] truthfulqa_mc2: {'acc,none': 0.388...
+Then, in your config, you only need to replace two fields: ``output_dir`` and ``checkpoint_files``. Notice
+that we are using the merged weights, and not the LoRA adapters.
 
-The model has an accuracy around 38.8%. Let's compare this with the fine-tuned model.
+.. code-block:: yaml
 
+    # TODO: update to your desired epoch
+    output_dir: /tmp/torchtune/llama3_2_3B/lora_single_device/epoch_0
```
Comment (felipemello1) [inline docs/source/tutorials/e2e_flow.rst:209]: if we do have epoch_0 and epoch_1, then maybe we should update all of the paths to epoch_1. I originally did epoch_0 to be on the safe side

File: docs/source/tutorials/e2e_flow.rst [hunk 6]
```diff
@@ -338,111 +355,155 @@ For Llama models, you can run generation directly in torchao on the quantized mo
...
-with interoperability with other libraries since torchtune doesn't add yet
-another format to the mix.
+Use with Hugging Face ``from_pretrained()``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-Let's take a look at an example of how this would work with a popular codebase
-used for running performant inference with LLMs -
-`gpt-fast <https://github.com/pytorch-labs/gpt-fast/tree/main>`_. This section
-assumes that you've cloned that repository on your machine.
+**Case 1: HF using BASE MODEL + trained adapter**
```
Comment (felipemello1) [inline docs/source/tutorials/e2e_flow.rst:368]: evan left some comments on this also in the original PR, avoiding the upper case

File: docs/source/tutorials/e2e_flow.rst [hunk 7]
```diff
@@ -25,222 +25,239 @@ from the community seemlessly with torchtune.
...
-    checkpointer.output_dir=<checkpoint_dir>
-
-
-Once training is complete, you'll see the following in the logs.
-
-.. code-block:: bash
-
-    [_checkpointer.py:473] Model checkpoint of size 9.98 GB saved to <checkpoint_dir>/hf_model_0001_0.pt
+We'll fine-tune using our
+`single device LoRA recipe <https://github.com/pytorch/torchtune/blob/main/recipes/lora_finetune_single_device.py>`_
```
Comment (SalmanMohammadi) [inline docs/source/tutorials/e2e_flow.rst:102]:
```suggestion
:ref:`single device LoRA recipe <lora_finetune_recipe_label>`
```

File: docs/source/tutorials/e2e_flow.rst [hunk 8]
```diff
@@ -25,222 +25,239 @@ from the community seemlessly with torchtune.
...
+3) **epoch_{}**: Contains your new trained model weights plus all original files of the model, except the checkpoints, making it easy for you to choose an specific epoch to run inference on or push to a model hub;
 
-The final trained weights are merged with the original model and split across two checkpoint files
-similar to the source checkpoints from the HF Hub
-(see the :ref:`LoRA Tutorial <lora_finetune_label>` for more details).
-In fact the keys will be identical between these checkpoints.
-We also have a third checkpoint file which is much smaller in size
-and contains the learnt LoRA adapter weights. For this tutorial, we'll only use the model
-checkpoints and not the adapter weights.
+.. code-block:: text
```
Comment (SalmanMohammadi) [inline docs/source/tutorials/e2e_flow.rst:136]:
```suggestion
.. code-block:: bash
```

File: docs/source/tutorials/e2e_flow.rst [hunk 9]
```diff
@@ -25,222 +25,239 @@ from the community seemlessly with torchtune.
...
+`single device LoRA recipe <https://github.com/pytorch/torchtune/blob/main/recipes/lora_finetune_single_device.py>`_
+and use the standard settings from the
+`default config <https://github.com/pytorch/torchtune/blob/main/recipes/configs/llama3_2/3B_lora_single_device.yaml>`_.
 
-    [_checkpointer.py:473] Model checkpoint of size 3.50 GB saved to <checkpoint_dir>/hf_model_0002_0.pt
+This will fine-tune our model using a ``batch_size=2`` and ``dtype=bfloat16``. With these settings the model
+should have a peak memory usage of ~16GB and total training time of around 2-3 hours for each epoch.
 
-    [_checkpointer.py:484] Adapter checkpoint of size 0.01 GB saved to <checkpoint_dir>/adapter_0.pt
+.. code-block:: text
```
Comment (SalmanMohammadi) [inline docs/source/tutorials/e2e_flow.rst:109]:
```suggestion
.. code-block:: bash
```

File: docs/source/tutorials/e2e_flow.rst [hunk 10]
```diff
@@ -25,222 +25,239 @@ from the community seemlessly with torchtune.
...
-.. code-block:: bash
-
-    tune run eleuther_eval --config ./custom_eval_config.yaml
+    # Quantization specific args
+    quantizer: null
 
+For this tutorial we'll use the `truthfulqa_mc2 <https://github.com/sylinrl/TruthfulQA>`_ task from the harness.
 
-The results should look something like this.
+This task measures a model's propensity to be truthful when answering questions and
```
Comment (SalmanMohammadi) [inline docs/source/tutorials/e2e_flow.rst:249]: nice use of the word propensity

File: docs/source/tutorials/e2e_flow.rst [hunk 11]
```diff
@@ -25,222 +25,239 @@ from the community seemlessly with torchtune.
...
+            GPU peak memory active: 6.21 GiB
+    Tokenizer is initialized from file.
+    Optimizer and loss are initialized.
+    Loss is initialized.
+    Dataset and Sampler are initialized.
+    Learning rate scheduler is initialized.
+    Profiling disabled.
+    Profiler config after instantiation: {'enabled': False}
+    1|3|Loss: 1.943998098373413:   0%|                    | 3/1617 [00:21<3:04:47,  6.87s/it]
+
```
Comment (felipemello1) [inline docs/source/tutorials/e2e_flow.rst:128]: we should make it a new section:

"Inspecting your outputs" or "Understanding the model outputs" or something like that. Looking at the docs, the finetuning section is too long: https://docs-preview.pytorch.org/pytorch/torchtune/2192/tutorials/e2e_flow.html

File: docs/source/tutorials/e2e_flow.rst [hunk 12]
```diff
@@ -25,222 +25,239 @@ from the community seemlessly with torchtune.
...
-    tune run eleuther_eval --config ./custom_eval_config.yaml
+    # Quantization specific args
+    quantizer: null
 
+For this tutorial we'll use the `truthfulqa_mc2 <https://github.com/sylinrl/TruthfulQA>`_ task from the harness.
 
-The results should look something like this.
+This task measures a model's propensity to be truthful when answering questions and
+measures the model's zero-shot accuracy on a question followed by one or more true
+responses and one or more false responses
```
Comment (SalmanMohammadi) [inline docs/source/tutorials/e2e_flow.rst:251]:
```suggestion
responses and one or more false responses.
```

**Review by pbontrager — APPROVED** [review]
This is a really great update to modernize this tutorial. I left some comments that you can pick and choose what you like.

File: docs/source/tutorials/e2e_flow.rst [hunk 13]
```diff
@@ -25,222 +25,239 @@ from the community seemlessly with torchtune.
...
 
-.. code-block:: bash
+.. code-block:: text
```
Comment (SalmanMohammadi) [inline docs/source/tutorials/e2e_flow.rst:253]: Maybe ignore my above comments, does bash render weirdly or something?

File: docs/source/tutorials/e2e_flow.rst (hunk 2 above)
Comment (felipemello1) [reply docs/source/tutorials/e2e_flow.rst:134]: skip one extra line after 3).
image

File: docs/source/tutorials/e2e_flow.rst [hunk 14]
```diff
@@ -25,222 +25,239 @@ from the community seemlessly with torchtune.
...
+    └── recipe_state
+        └── recipe_state.pt
+
+Let's understand the files:
+
+- `adapter_model.safetensors` and `adapter_model.pt` are your LoRA trained adapter weights. We save a duplicated .pt version of it to facilitate resuming from checkpoint.
+- `ft-model-{}-of-{}.safetensors` are your trained full model weights (not adapters). When LoRA finetuning, these are only present if we set ``save_adapter_weights_only=False``. In that case, we merge the merged base model with trained adapters, making inference easier.
+- `adapter_config.json` is used by Huggingface PEFT when loading an adapter (more on that later);
+- `model.safetensors.index.json` is used by Huggingface .from_pretrained when loading the model weights (more on that later)
+- All other files were originally in the checkpoint_dir. They are automatically copied during training. Files over 100MiB and ending on .safetensors, .pth, .pt, .bin are ignored, making it lightweight.
```
Comment (felipemello1) [inline docs/source/tutorials/e2e_flow.rst:171]: should we  add double quotations here for better rendering?

image

File: docs/source/tutorials/e2e_flow.rst [hunk 15]
```diff
@@ -25,222 +25,239 @@ from the community seemlessly with torchtune.
...
-Overview
---------
+Finetune your model
+-------------------
 
-Fine-tuning an LLM is usually only one step in a larger workflow. An example workflow that you
-might have can look something like this:
+First, let's download a model using the tune CLI. The following command will download the `Llama3.2 3B Instruct <https://ai.meta.com/blog/llama-3-2-connect-2024-vision-edge-mobile-devices/>`_
+model from the Hugging Face Hub and save it the local filesystem. Hugging Face uploaded the original
+weights (``consolidated.00.pth``) and the weights compatible with the `from_pretrained() <https://huggingface.co/docs/huggingface_hub/main/en/guides/integrations#frompretrained>`_ API (`*.safetensors`).
```
Comment (ebsmothers) [inline docs/source/tutorials/e2e_flow.rst:33]: Not sure if you want *.safetensors to render as code, but single backticks render as italics FYI

File: docs/source/tutorials/e2e_flow.rst [hunk 16]
```diff
@@ -250,76 +267,76 @@ For this, we'll use the
...
 
+    $ tune run generate --config ./custom_generation_config.yaml prompt="tell me a joke. "
+    Tell me a joke. Here's a joke for you:
 
-Indeed, the bridge is pretty cool! Seems like our LLM knows a little something about the
-Bay Area!
+    What do you call a fake noodle?
 
-|
+    An impasta!
```
Comment (SalmanMohammadi) [inline docs/source/tutorials/e2e_flow.rst:336]: lmaooooooo

File: docs/source/tutorials/e2e_flow.rst [hunk 17]
```diff
@@ -25,222 +25,239 @@ from the community seemlessly with torchtune.
...
-.. code-block:: bash
-
-    tune run eleuther_eval --config ./custom_eval_config.yaml
-    checkpointer.checkpoint_dir=<checkpoint_dir> \
-    tokenizer.path=<checkpoint_dir>/tokenizer.model
-
-    [evaluator.py:324] Running loglikelihood requests
-    [eleuther_eval.py:195] Eval completed in 121.27 seconds.
-    [eleuther_eval.py:197] truthfulqa_mc2: {'acc,none': 0.388...
+Then, in your config, you only need to replace two fields: ``output_dir`` and ``checkpoint_files``. Notice
```
Comment (felipemello1) [inline docs/source/tutorials/e2e_flow.rst:203]: i guess this is wrong, the user also has to update model, model_type and tokenizer

File: docs/source/tutorials/e2e_flow.rst (hunk 16 above)
Comment (pbontrager) [reply docs/source/tutorials/e2e_flow.rst:336]: You're just trying to make sure the LLMs like you

File: docs/source/tutorials/e2e_flow.rst [hunk 18]
```diff
@@ -25,222 +25,239 @@ from the community seemlessly with torchtune.
         :ref:`checkpoints <understand_checkpointer>`
 
 
-Overview
---------
+Finetune your model
```
[the remark this answers is not in the document]
Comment (joecummings) [reply docs/source/tutorials/e2e_flow.rst:28]: IMO, no

File: docs/source/tutorials/e2e_flow.rst [hunk 19]
```diff
@@ -250,76 +267,76 @@ For this, we'll use the
...
-.. code-block:: bash
+.. code-block:: text
 
-    tune cp generation ./custom_generation_config.yaml
+    $ tune cp generation ./custom_generation_config.yaml
+    Copied file to custom_generation_config.yaml
 
-Let's modify ``custom_generation_config.yaml`` to include the following changes.
+Let's modify ``custom_generation_config.yaml`` to include the following changes. Again, you only need
+ to replace two fields: ``output_dir`` and ``checkpoint_files``
```
Comment (felipemello1) [inline docs/source/tutorials/e2e_flow.rst:278]: here too, this is only true is the correct model is already defined

File: docs/source/tutorials/e2e_flow.rst [hunk 20]
```diff
@@ -25,222 +25,239 @@ from the community seemlessly with torchtune.
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
 .. TODO (SalmanMohammadi) ref eval recipe docs
 
 torchtune integrates with
 `EleutherAI's evaluation harness <https://github.com/EleutherAI/lm-evaluation-harness>`_.
 An example of this is available through the
-``eleuther_eval`` recipe. In this tutorial, we're going to directly use this recipe by
-modifying its associated config ``eleuther_evaluation.yaml``.
+`eleuther_eval <https://github.com/pytorch/torchtune/blob/main/recipes/eleuther_eval.py>`_ recipe. In this tutorial, we're going to directly use this recipe by
```
[the remark this answers is not in the document]
Comment (joecummings) [reply docs/source/tutorials/e2e_flow.rst:188]: It's inflight... see #2186 

Do you want to land it?

File: docs/source/tutorials/e2e_flow.rst [hunk 21]
```diff
@@ -250,76 +267,76 @@ For this, we'll use the
...
-Indeed, the bridge is pretty cool! Seems like our LLM knows a little something about the
-Bay Area!
+    What do you call a fake noodle?
 
-|
+    An impasta!
 
-Speeding up Generation using Quantization
------------------------------------------
+Introduce some quantization
```
Comment (felipemello1) [inline docs/source/tutorials/e2e_flow.rst:338]: our quantize recipe only outputs 1 file that is .pt. I am tempted to say that we should remove this from here until we have better support for quantization, e.g. output safetensors and multiple checkpoints. Any thoughts?

File: docs/source/tutorials/e2e_flow.rst (hunk 13 above)
Comment (joecummings) [reply docs/source/tutorials/e2e_flow.rst:255]: Yeah when you display the outputs from bash.

File: docs/source/tutorials/e2e_flow.rst [hunk 22]
```diff
@@ -25,222 +25,239 @@ from the community seemlessly with torchtune.
...
-Our fine-tuned model gets ~48% on this task, which is ~10 points
-better than the baseline. Great! Seems like our fine-tuning helped.
-
-|
+    ...
 
-Generation
------------
+Generate some output
+~~~~~~~~~~~~~~~~~~~~
```
[the remark this answers is not in the document]
Comment (joecummings) [reply docs/source/tutorials/e2e_flow.rst:262]: Addressed in different channel.

**Pull request merged**